src/db.rs

```diff
@@ -170,19 +171,21 @@ impl DbInner {
             };
             let mut iter = SstIterator::new(&sst, self.table_store.clone(), 1, 1);
             // iterate over the WAL SSTs in reverse order to ensure we recover in write-order
-            while let Some(kv) = iter.next_entry().await? {
-                // TODO: it's not ideal that we have to take this lock for every kv. We can solve
-                //       this by either:
-                //       1. detaching this method from self and calling it before initializing
-                //          DbInner. The downside is we can't use member methods of DbInner
-                //          like maybe_freeze_wal
-                //       2. accumulating kv-pairs in memory and bulk-applying the writes
+            {
+                // buffer the WAL entries to bulk replay them into the memtable, this buffer is reused
+                // across replaying different WAL SSTs.
+                wal_replay_buf.clear();
```

Comment:
I'd prefer it if we just create a brand new buffer every time rather than clearing the old one. Makes it less likely that we'd introduce a bug where we don't clear it for some reason and wind up writing bad data.

Reply:
@rodesai makes sense, i've fixed it, PTAL